scipy/spatial/ckdtree/src/ckdtree_cpp_methods.h

```diff
@@ -0,0 +1,107 @@
+
+#ifndef CKDTREE_CPP_METHODS
+#define CKDTREE_CPP_METHODS
+
+#ifdef CKDTREE_METHODS_IMPL
+#define CKDTREE_EXTERN extern "C"
+#else
+#define CKDTREE_EXTERN extern "C" 
+struct ckdtree;
+#endif 
+
+extern npy_float64 infinity;
+extern int number_of_processors;
+
+#ifndef NPY_LIKELY
+#define NPY_LIKELY(x) (x)
+#endif
+
+#include <cmath>
+
+// Utility functions
+// =================
+
+inline npy_float64 
+dmax(const npy_float64 x, const npy_float64 y) 
+{
+    if (x > y)
+        return x;
+    else
+        return y;
+};
+
+inline npy_float64 
+dabs(const npy_float64 x)
+{
+    if (x > 0)
+        return x;
+    else
+        return -x;
+}
+
+
+// Measuring distances
+// ===================
+
+inline npy_float64 
+_distance_p(const npy_float64 *x, const npy_float64 *y,
+            const npy_float64 p, const npy_intp k,
+            const npy_float64 upperbound)
+{    
+    // Compute the distance between x and y
+    //
+    // Computes the Minkowski p-distance to the power p between two points.
+    // If the distance**p is larger than upperbound, then any number larger
+    // than upperbound may be returned (the calculation is truncated).
```

Comment:
C++ is an abomination, but the closest fit for the Cython code in cKDTree. Otherwise I wouldn't have considered to use it.